json_rpc/private/jrpc_sys.nim

```diff
@@ -0,0 +1,353 @@
+      str*: string
+    of riNull:
+      discard
+
+  # Request received by server
+  RequestRx* = object
+    jsonrpc* : results.Opt[JsonRPC2]
+    id*      : RequestId
+    `method`*: results.Opt[string]
+    params*  : RequestParamsRx
+
+  # Request sent by client
+  RequestTx* = object
+    jsonrpc* : JsonRPC2
+    id*      : results.Opt[RequestId]
+    `method`*: string
+    params*  : RequestParamsTx
+
+  ResponseError* = object
+    code*   : int
+    message*: string
+    data*   : results.Opt[JsonString]
+
+  ResponseKind* = enum
+    rkResult
+    rkError
+
+  # Response sent by server
+  ResponseTx* = object
+    jsonrpc*  : JsonRPC2
+    id*       : RequestId
+    case kind*: ResponseKind
+    of rkResult:
+      result* : JsonString
+    of rkError:
+      error*  : ResponseError
+
+  # Response received by client
+  ResponseRx* = object
+    jsonrpc*: results.Opt[JsonRPC2]
+    id*     : results.Opt[RequestId]
+    result* : results.Opt[JsonString]
+    error*  : results.Opt[ResponseError]
+
+  ReBatchKind* = enum
+    rbkSingle
+    rbkMany
+
+  RequestBatchRx* = object
+    case kind*: ReBatchKind
+    of rbkMany:
+      many*  : seq[RequestRx]
+    of rbkSingle:
+      single*: RequestRx
+
+  RequestBatchTx* = object
+    case kind*: ReBatchKind
+    of rbkMany:
+      many*  : seq[RequestTx]
+    of rbkSingle:
+      single*: RequestTx
+
+  ResponseBatchRx* = object
+    case kind*: ReBatchKind
+    of rbkMany:
+      many*  : seq[ResponseRx]
+    of rbkSingle:
+      single*: ResponseRx
+
+  ResponseBatchTx* = object
+    case kind*: ReBatchKind
+    of rbkMany:
+      many*  : seq[ResponseTx]
+    of rbkSingle:
+      single*: ResponseTx
+
+# don't mix the json-rpc system encoding with the
+# actual response/params encoding
+createJsonFlavor JrpcSys,
+  requireAllFields = false
+
+const
+  JsonRPC2Literal = JsonString("\"2.0\"")
+
+{.push gcsafe, raises: [].}
+
+func hash*(x: RequestId): hashes.Hash =
+  var h = 0.Hash
+  case x.kind:
+  of riNumber: h = h !& hash(x.num)
+  of riString: h = h !& hash(x.str)
+  of riNull: h = h !& hash("null")
+  result = !$(h)
+
+func `$`*(x: RequestId): string =
+  case x.kind:
+  of riNumber: $x.num
+  of riString: x.str
+  of riNull: "null"
+
+func `==`*(a, b: RequestId): bool =
+  if a.kind != b.kind:
+    return false
+  case a.kind
+  of riNumber: a.num == b.num
+  of riString: a.str == b.str
+  of riNull: true
+
+func meth*(rx: RequestRx): Opt[string] =
+  rx.`method`
+
+proc readValue*(r: var JsonReader[JrpcSys], val: var JsonRPC2)
+      {.gcsafe, raises: [IOError, JsonReaderError].} =
+  let version = r.parseAsString()
+  if version != JsonRPC2Literal:
+    r.raiseUnexpectedValue("Invalid JSON-RPC version, want=" &
+      JsonRPC2Literal.string & " got=" & version.string)
+
+proc writeValue*(w: var JsonWriter[JrpcSys], val: JsonRPC2)
+      {.gcsafe, raises: [IOError].} =
+  w.writeValue JsonRPC2Literal
+
+proc readValue*(r: var JsonReader[JrpcSys], val: var ResponseError)
```

Comment:
if we disable automatic object serialization in JrcpSys and then opt in for these few types, I think we can avoid these manual case implementations?

Reply:
Right, it's too easy to write custom serializer using the new json-serialization, make me forget to use new features in nim-serialization. Now it's fixed.

Reply:
:rofl: